### ISSUE TITLE
SmartOS USB Boot hangs on ProLiant DL380 G10

### ISSUE THREAD

**Issue by akole** [joyent/smartos-live#824]
Using 20190117T003704Z as the USB image, boot hangs after "installing rts, module id 156."

@rmustacc suggested on IRC to get some data from kmdb for help in debugging the issue which can be found here: [kmdb output](https://www.dropbox.com/s/o99q4btf7kvxzeb/kmdb_data.txt?dl=0)

edit: Just to add, I am using the Triton version of the image here.

**Comment by rmustacc** [comment]
On 1/24/19 14:13 , Andy wrote:
> Using 20190117T003704Z as the USB image, boot hangs after "installing rts, module id 156."
> 
> @rmustacc suggested on IRC to get some data from kmdb for help in debugging the issue which can be found here: [kmdb output](https://www.dropbox.com/s/o99q4btf7kvxzeb/kmdb_data.txt?dl=0)
> 
> 

Thanks for providing this information. This helps get me pointed in the
right direction. What I've noticed from looking at this is that we have
a bunch of different stacks that are blocked behind various issues in
/devices. For example:

```
fffffd67c5502440 SLEEP    CV                      1
                 swtch+0x141
                 cv_wait+0x70
                 ndi_devi_enter+0x7f
                 devi_config_one+0x3be
                 ndi_devi_config_one+0xa8
                 devfs`dv_find+0x23b
                 devfs`devfs_lookup+0x37
                 fop_lookup+0xa3
                 lookuppnvp+0x230
                 lookuppnatcred+0x176
                 lookupnameatcred+0xdd
                 lookupnameat+0x39
                 vn_openat+0x315
                 copen+0x204
                 openat32+0x27
                 open32+0x25
                 _sys_sysenter_post_swapgs+0x153

fffff946057c8c20 SLEEP    CV                      1
                 swtch+0x141
                 cv_wait+0x70
                 ndi_devi_enter+0x7f
                 i_ddi_attach_node_hierarchy+0x5a
                 attach_driver_nodes+0xb2
                 ddi_hold_installed_driver+0x130
                 viona`set_viona_tx_mode+0x26
                 viona`viona_attach+0xa9
                 devi_attach+0x92
                 attach_node+0xa7
                 i_ndi_config_node+0x7d
                 i_ddi_attachchild+0x48
                 devi_attach_node+0x5e
                 config_immediate_children+0xbf
                 devi_config_common+0xd9
                 mt_config_thread+0x58
                 thread_start+8

fffff94605818c20 SLEEP    CV                      1
                 swtch+0x141
                 cv_wait+0x70
                 ndi_devi_enter+0x7f
                 i_ddi_attach_node_hierarchy+0x5a
                 attach_driver_nodes+0xb2
                 i_ddi_attach_pseudo_node+0x2f
                 softmac`softmac_create+0x46
                 net_dacf`net_postattach+0x3e
                 dacf_op_invoke+0x202
                 dacf_process_rsrvs+0xdb
                 dacfc_postattach+0x2d
                 postattach_node+0x44
                 i_ndi_config_node+0x148
                 i_ddi_attachchild+0x48
                 devi_attach_node+0x5e
                 config_immediate_children+0xbf
                 devi_config_common+0xd9
                 mt_config_thread+0x58
                 thread_start+8
```

Now, the question will be what is the thread that holds these. The way I
normally do this is to take one of these threads and run <thread
address>::findstack -v. From that, I take the first argument to
ndi_devi_enter() and print it out as:

<addr>::print struct dev_info devi_busy_thread | ::findstack -v

That itself may be blocked on something else, so I then repeat the
procedure. I can talk through how one might step through this
interactively if that helps.

Robert

**Comment by rmustacc** [comment]
On 1/24/19 16:24 , Andy wrote:
> Here is what is shows:
> 
> ...

Thanks, that was really helpful. I think I have a lead on what's going
on. Can you confirm what NICs you have in that system?

**Comment by akole** [comment]
At the moment there is two, both of them integrated if I'm not mistaken.

The first one is: HP FlexFabric 10Gb 2port 534FLR-SFP+ Adapter (Embedded ALOM)
The second is: HPE Ethernet 1Gb 4-port 331i Adapter (Embedded LOM)

If I remember correctly, at least the 1Gb one was with a Broadcom chip. We also had one FC adapter and another NIC with Intel chipset as cards but those were removed when we started debugging this problem.

**Comment by rmustacc** [comment]
On 1/24/19 16:59 , Andy wrote:
> At the moment there is two, both of them integrated if I'm not mistaken.
> 
> The first one is: HP FlexFabric 10Gb 2port 534FLR-SFP+ Adapter (Embedded ALOM)
> The second is: HPE Ethernet 1Gb 4-port 331i Adapter (Embedded LOM)
> 
> If I remember correctly, at least the 1Gb one was with a Broadcom chip. We also had one FC adapter and another NIC with Intel chipset as cards but those were removed when we started debugging this problem.
> 
> 

OK, that helps confirm my theory. I'll have a full write up a bit later.
At least, I think I understand the problem at this point.

**Comment by rmustacc** [comment]
I've filed OS-7538 (https://smartos.org/bugview/OS-7538) to cover this.
 A copy of the meat of the write up is below:

There are two main threads that we care about. The first is the viona
thread:

[0]> fffff946057c8c20::findstack -v
stack pointer for thread fffff946057c8c20 (mt_config_thread()):
fffff946057c8790
[ fffff946057c8790 _resume_from_idle+0x12b() ]
  fffff946057c87c0 swtch+0x141()
  fffff946057c8800 cv_wait+0x70(fffffd66d881b954, fffffd66d881b868)
  fffff946057c8840 ndi_devi_enter+0x7f(fffffd66d881b800, fffff946057c886c)
  fffff946057c8890 i_ddi_attach_node_hierarchy+0x5a(fffffd66d880ed58)
  fffff946057c88d0 attach_driver_nodes+0xb2(5d)
  fffff946057c8920 ddi_hold_installed_driver+0x130(5d)
  fffff946057c8940 viona`set_viona_tx_mode+0x26()
  fffff946057c8970 viona`viona_attach+0xa9(fffffd67c4165560, 0)
  fffff946057c89e0 devi_attach+0x92(fffffd67c4165560, 0)
  fffff946057c8a20 attach_node+0xa7(fffffd67c4165560)
  fffff946057c8a70 i_ndi_config_node+0x7d(fffffd67c4165560, 6, 0)
  fffff946057c8aa0 i_ddi_attachchild+0x48(fffffd67c4165560)
  fffff946057c8ae0 devi_attach_node+0x5e(fffffd67c4165560, 4004048)
  fffff946057c8b60 config_immediate_children+0xbf(fffffd66d9ba0d58,
4004048, ffffffff)
  fffff946057c8bb0 devi_config_common+0xd9(fffffd66d9ba0d58, 4004048,
ffffffff)
  fffff946057c8c00 mt_config_thread+0x58(fffffd66ff32f210)
  fffff946057c8c10 thread_start+8()


The second is the softmac thread:

[0]> fffff94605818c20::findstack -v
stack pointer for thread fffff94605818c20 (mt_config_thread()):
fffff94605818580
[ fffff94605818580 _resume_from_idle+0x12b() ]
  fffff946058185b0 swtch+0x141()
  fffff946058185f0 cv_wait+0x70(fffffd66d9ba0eac, fffffd66d9ba0dc0)
  fffff94605818630 ndi_devi_enter+0x7f(fffffd66d9ba0d58, fffff9460581865c)
  fffff94605818680 i_ddi_attach_node_hierarchy+0x5a(fffffd66d9b8e7f8)
  fffff946058186c0 attach_driver_nodes+0xb2(38)
  fffff94605818700 i_ddi_attach_pseudo_node+0x2f(fffffffff801cb97)
  fffff94605818870 softmac`softmac_create+0x46(fffffd66d8811008, 5d00000001)
  fffff946058188c0 net_dacf`net_postattach+0x3e(fffffd67c5504298, 0, 0)
  fffff94605818950 dacf_op_invoke+0x202(fffffd664d68dde8,
fffffd67c5504298, 0)
  fffff946058189c0 dacf_process_rsrvs+0xdb(fffffd66d8811168, 1, 1)
  fffff946058189f0 dacfc_postattach+0x2d(fffffd66d8811008)
  fffff94605818a20 postattach_node+0x44(fffffd66d8811008)
  fffff94605818a70 i_ndi_config_node+0x148(fffffd66d8811008, 6, 0)
  fffff94605818aa0 i_ddi_attachchild+0x48(fffffd66d8811008)
  fffff94605818ae0 devi_attach_node+0x5e(fffffd66d8811008, 4004048)
  fffff94605818b60 config_immediate_children+0xbf(fffffd66d881b800,
4004048, ffffffff)
  fffff94605818bb0 devi_config_common+0xd9(fffffd66d881b800, 4004048,
ffffffff)
  fffff94605818c00 mt_config_thread+0x58(fffffd67c49b7730)
  fffff94605818c10 thread_start+8()


Let's first talk about what each thread is trying to logically do, and
then we'll come back to why we're in this situation and how that lead to
a system being corked.

The viona thread is in the process of attaching. It is trying to check
if the bnxe driver is attached to work around an issue in the driver.

The softmac thread is following up after a GLDv3 driver has been
attached to attach an instance of the softmac pseudo device driver for
that device.

Critically, because the viona thread is in the middle of viona_attach(),
its parent, the psuedo nexus driver has its dev_info structure held. It
it is trying to get information about bnxe being attached. This first
walks all of the parents of bnxe to see if they're attached. The parent
we're on is the pcie root complex that has as its child, almost
certainly, an instance of bnxe. However, this parent is owned by the
softmac thread.

The softmac thread has just finished attaching an instance of bnxe.
bnxe's parent is held, which is the pcie instance that the viona thread
wants. Now, because we finished attaching bnxe, we want to attach a
pseudo node. To do this, we need to start with the pseudo nexus driver,
which is the one that viona has.

So, to summarize the deadlock:

* viona has pseudo, wants pcie root complex
* softmac has pcie root complex, wants pseudo

Fundamentally, this is an unsafe combination. It's more likely that
viona is in the wrong here by being in its explicit attach and trying to
call into and lock the devices tree. However, it's not clear to me that
the dacf framework that softmac is leveraging is really safe either.
It's existence kind of bars the other way from working, which I don't
think has come up very often.

In terms of fixes, there are a couple of ways that we can probably proceed:

1) Patrick Mooney had the idea that we could safely move the check that
viona is doing from attach(9E) to mc_start(9E) context. mc_start(9E)
should never run in attach context and thus not have holds on the
devices tree; however, we should confirm this.

2) We should figure out why viona is special casing bnxe and just fix
bnxe to no longer need such special casing. This would allow us to
remove the logic entirely.

**Comment by rmustacc** [comment]
On 1/25/19 2:26 , Andy wrote:
> Great! Thank you for taking the time looking into this.
> 
> Is there any instant remedy I could try while waiting for fix? 
> Disabling bnxe on boot probably would get the initial install going but I wonder if there's anything else that could let us keep the 10 gig going as well.
> 

The only initial thing I can think of is disabling bnxe on boot or via
the bios. However, we are working on a fix to this and probably can have
something for you to try early next week.

One other issue with the HP G10 systems is the HBA support. We don't
have much support for the g10 controllers. Just wanted to throw that out
in advance so as to minimize the surprise, sorry.

**Comment by pfmooney** [comment]
@akole I've drafted up a fix to this issue (https://cr.joyent.us/#/c/5431/).  If you are interested in testing it on the machine of yours which is experiencing the issue, I'd be happy to provide you with a platform image (or USB image).

**Comment by akole** [comment]
@pfmooney I'd be happy to test the fix and a USB-image would probably be the best option for now.
Any data I need to collect when booting with the new image?

**Comment by pfmooney** [comment]
I shouldn't need anything beyond confirming that the boot-up process progresses further than you were seeing before.  I've uploaded a SmartOS (sans Triton) image to test with [here](http://us-east.manta.joyent.com/patrick.mooney/public/pi/platform.OS-7538.usb.bz2).  I would hope and expect it's able to boot to the point of starting the installer.

Please let me know if you have questions/comments.

**Comment by akole** [comment]
Tested the image and everything works now. No more hang and bnxe shows up as it should.

**Comment by rmustacc** [comment]
On 1/29/19 4:17 , Andy wrote:
> Tested the image and everything works now. No more hang and bnxe shows up as it should.
> 

Great, that's good to hear.

**Comment by pfmooney** [comment]
The fix for this is merged in joyent/illumos-joyent@4bf87f0f9a9dec4e46764fe7c2c7c342fe2f32d9.
Thanks again for reporting this and helping with the testing.
